### ISSUE TITLE
Add d3.layout.hierarchy.links for easier custom layouts.

### ISSUE THREAD

**Issue by hooverlunch** [d3/d3#2421]
Hi there. I'm trying to write my own layout engine similar to a tree layout. I'm finding that I need to copy and paste this:

```
// A method assignment helper for hierarchy subclasses.
function d3_layout_hierarchyRebind(object, hierarchy) {
  d3.rebind(object, hierarchy, "sort", "children", "value");

  // Add an alias for nodes and links, for convenience.
  object.nodes = object;
  object.links = d3_layout_hierarchyLinks;

  return object;
}

// Returns an array source+target objects for the specified nodes.
function d3_layout_hierarchyLinks(nodes) {
  return d3.merge(nodes.map(function(parent) {
    return (parent.children || []).map(function(child) {
      return {source: parent, target: child};
    });
  }));
}
```

into my JS or it won't run. Am I doing something wrong or is writing custom layouts not supported?

Thanks!

**Comment by jasondavies** [comment]
These are private helper functions used in various D3 hierarchical layouts.  If you are referencing them from outside D3 then yes, you’ll need to make a copy.  I suppose we could potentially make things a bit easier if the links method (d3_layout_hierarchyLinks) was assigned to d3.layout.hierarchy instances by default.  That way, you would only need to rebind any methods that you were interested in.

**Comment by jasondavies** [comment]
For the moment, you can cheat and avoid copying d3_layout_hierarchyLinks by doing something like:

```js
function myLayout() {
  function layout() { … }

  layout.links = d3.layout.tree().links;

  return layout;
}
```

**Comment by hooverlunch** [comment]
Thanks! For completeness/convention, I also had to add `tree.nodes = tree;`.

**Comment by mbostock** [comment]
This is fixed in D3 4.0 with the new [*node*.links](https://github.com/d3/d3-hierarchy#node_links) method on [d3.hierarchy](https://github.com/d3/d3-hierarchy#hierarchy).

**Comment by hooverlunch** [comment]
Thanks!!!
